mlflow/store/artifact/optimized_s3_artifact_repo.py

```diff
@@ -0,0 +1,275 @@
+import json
+import logging
+import math
+import os
+import posixpath
+import urllib.parse
+from mimetypes import guess_type
+
+from mlflow.entities import FileInfo
+from mlflow.environment_variables import (
+    MLFLOW_ENABLE_MULTIPART_UPLOAD,
+    MLFLOW_S3_UPLOAD_EXTRA_ARGS,
+)
+from mlflow.exceptions import MlflowException
+from mlflow.protos.databricks_artifacts_pb2 import ArtifactCredentialInfo
+from mlflow.store.artifact.cloud_artifact_repo import (
+    _MULTIPART_UPLOAD_CHUNK_SIZE,
+    CloudArtifactRepository,
+    _complete_futures,
+)
+from mlflow.store.artifact.s3_artifact_repo import _get_s3_client
+from mlflow.utils.file_utils import read_chunk
+from mlflow.utils.request_utils import cloud_storage_http_request
+from mlflow.utils.rest_utils import augmented_raise_for_status
+
+_logger = logging.getLogger(__name__)
+
+
+class OptimizedS3ArtifactRepository(CloudArtifactRepository):
+    """Stores artifacts on Amazon S3."""
+
+    def __init__(
+        self,
+        artifact_uri,
+        access_key_id=None,
+        secret_access_key=None,
+        session_token=None,
+        addressing_style="path",
+        s3_endpoint_url=None,
+    ):
+        super().__init__(artifact_uri)
+        self._access_key_id = access_key_id
+        self._secret_access_key = secret_access_key
+        self._session_token = session_token
+        self._addressing_style = addressing_style
+        self._s3_endpoint_url = s3_endpoint_url
+        self.bucket, self.bucket_path = self.parse_s3_compliant_uri(self.artifact_uri)
+        self._region_name = self._get_region_name()
+
+    def _get_region_name(self):
+        temp_client = _get_s3_client(
+            addressing_style=self._addressing_style,
+            access_key_id=self._access_key_id,
+            secret_access_key=self._secret_access_key,
+            session_token=self._session_token,
+            s3_endpoint_url=self._s3_endpoint_url,
+        )
+        return temp_client.get_bucket_location(Bucket=self.bucket)["LocationConstraint"]
+
+    def _get_s3_client(self):
+        return _get_s3_client(
+            addressing_style=self._addressing_style,
+            access_key_id=self._access_key_id,
+            secret_access_key=self._secret_access_key,
+            session_token=self._session_token,
+            region_name=self._region_name,
+            s3_endpoint_url=self._s3_endpoint_url,
+        )
+
+    def parse_s3_compliant_uri(self, uri):
```

Comment:
Another followup... this is a static method. It's also a utility. When refactoring, we should externalize this to a common util for both standard and optimized implementations.

Reply:
We recently changed this from a utility into a class method is so that the `R2ArtifactRepository` subclass can override this.